common/sampling.cpp

```diff
@@ -139,14 +139,23 @@ std::string common_sampler_params::print() const {
     return std::string(result);
 }
 
+bool common_sampler_trigger_grammar(const struct llama_model * model, common_sampler * gsmpl, const std::string & trigger) {
+    if (!llama_sampler_is_grammar_empty(gsmpl->grmr)) {
+        return false;
+    }
+    gsmpl->grmr   = llama_sampler_init_grammar(model, gsmpl->params.grammar.c_str(), "root");
+    llama_sampler_accept_str(gsmpl->grmr, trigger.c_str());
+    return true;
```

Comment:
If this is the only use case of `llama_sampler_accept_str()` maybe we can simply extend `llama_sampler_init_grammar()` to accept an initial string (i.e. the trigger) and avoid extending the API.

Reply:
Oh, you mean to move the trigger logic in the grammar itself?

Technically the triggers are even already in the grammar. It would all be *neater* if we could write `root ::= .*? ("trigger1" ... | "trigger2" ...)` (e.g. treat the head literal that comes after a reluctant `.*` as a trigger; we could even have the grammar pause itself / allow temporarily unconstrained sequences, e.g. `root ::= prefix .*? suffix` if we collect the triggers = head literals recursively in the `suffix` subgraph).

Reply:
> Oh, you mean to move the trigger logic in the grammar itself?

No, I had in mind something more like this:

```c
// llama.h
    LLAMA_API struct llama_sampler * llama_sampler_init_grammar(
            const struct llama_model * model,
                          const char * grammar_str,
                          const char * grammar_root,
                          const char * init_str); // optionally used to feed initial data to the grammar state

// common/sampling.cpp
    gsmpl->grmr  = llama_sampler_init_grammar(model, gsmpl->params.grammar.c_str(), "root", trigger.c_str());
    
    return true;
```

The logic for triggering remains in user code.